presto-main/src/main/java/com/facebook/presto/operator/OptimizedPartitionedOutputOperator.java

```diff
@@ -0,0 +1,757 @@
+        {
+            if (positions.length < positionCount) {
+                positions = new int[max(2 * positions.length, positionCount)];
+            }
+
+            this.positionCount = 0;
+        }
+
+        public void appendPosition(int position)
+        {
+            positions[positionCount++] = position;
+        }
+
+        public void appendData(DecodedBlockNode[] decodedBlocks, int fixedWidthRowSize, List<Integer> variableWidthChannels, OutputBuffer outputBuffer)
+        {
+            checkArgument(decodedBlocks.length == channelCount, "Unexpected number of blocks");
+
+            if (positionCount == 0) {
+                return;
+            }
+
+            if (channelCount == 0) {
+                bufferedRowCount += positionCount;
+                return;
+            }
+
+            initializeBlockEncodingBuffers(decodedBlocks);
+
+            for (int i = 0; i < channelCount; i++) {
+                blockEncodingBuffers[i].setupDecodedBlocksAndPositions(decodedBlocks[i], positions, positionCount);
+            }
+
+            calculateRowSizes(fixedWidthRowSize, variableWidthChannels);
+
+            int offset = 0;
+            do {
+                int batchSize = calculateNextBatchSize(fixedWidthRowSize, variableWidthChannels, offset);
+
+                for (int i = 0; i < channelCount; i++) {
+                    blockEncodingBuffers[i].setNextBatch(offset, batchSize);
+                    blockEncodingBuffers[i].copyValues();
+                }
+
+                bufferedRowCount += batchSize;
+                offset += batchSize;
+
+                if (bufferFull) {
+                    flush(outputBuffer);
+                    bufferFull = false;
+                }
+            }
+            while (offset < positionCount);
+        }
+
+        public long getInitialRetainedSizeInBytes()
+        {
+            return INSTANCE_SIZE + sizeOf(positions) + sizeOf(rowSizes);
+        }
+
+        public long getRetainedSizeInBytes()
+        {
+            long size = INSTANCE_SIZE + sizeOf(positions) + sizeOf(rowSizes);
+
+            if (blockEncodingBuffers != null) {
+                for (int i = 0; i < channelCount; i++) {
+                    size += blockEncodingBuffers[i].getRetainedSizeInBytes();
+                }
+            }
+
+            return size;
+        }
+
+        private void initializeBlockEncodingBuffers(DecodedBlockNode[] decodedBlocks)
+        {
+            // Create buffers has to be done after seeing the first page.
+            if (blockEncodingBuffers == null) {
+                blockEncodingBuffers = new BlockEncodingBuffers[channelCount];
+                for (int i = 0; i < channelCount; i++) {
+                    blockEncodingBuffers[i] = createBlockEncodingBuffers(decodedBlocks[i]);
+                }
+            }
+        }
+
+        private void calculateRowSizes(int fixedWidthRowSize, List<Integer> variableWidthChannels)
```

Comment:
calculateRowSizes is only called once for each incoming page and the result of it is the rowSizes for all blocks in the page. Its internal implementation calls accumulateRowSizes() for each blocks. So I think calculateXXX is better?

Reply:
@yingsu00 : Usually when I see a method named `calculateXXX`, I expected it to return a long or something similar, and don't update its object.

That being said, I do note there are some `calculateXXX` method in the codebase doesn't return a value but update the object (e.g. `DictionaryBlock#calculateCompactSize`). 

So, again a personal preference and take your judgement call then :)

Reply:
Still I think we can call it `calculateSerializedRowSizes` -- perhaps we should use `serializedRowSize` everywhere since it's easier to understand than `rowSize`

Reply:
@wenleix ah I see. How about naming it to setupSerializedRowSizes()? I'll also use serializedRowSize everywhere else.